### ISSUE TITLE
Relax certain validity requirements for locale strings

### ISSUE THREAD

**Issue by sffc** [tc39/ecma402#301]
Comment from @macchiati (Mark Davis, president of the Unicode Consortium):

> > Read the bullet point right after that:
> > > [[SortLocaleData]] and [[SearchLocaleData]] (for Intl.Collator) and [[LocaleData]] (for Intl.NumberFormat, Intl.DateTimeFormat, and Intl.PluralRules) are records that have fields for each locale contained in [[AvailableLocales]]. The value of each of these fields must be a record that has fields for each key contained in [[RelevantExtensionKeys]]. The value of each of these fields must be a non-empty list of those values defined in Unicode Technical Standard 35 for the given key that are supported by the implementation for the given locale, with the first element providing the default value.
> 
> I think this is dangerous. Validity checking should be an option — not a requirement within other services like matching.
> 
> Why? Because if you bake validity checking into code that could be running on out-of-date software (eg older version of ICU), then you will reject locales that are perfectly valid on an up-to-date system. For general-purpose software that is not absolutely guaranteed to be up-to-date, it is much better to reject ill-formed locales, but allow invalid ones — and ignore tags that you don't understand. And I don't think we can ever assume that JS implementations will always be up-to-date with the latest CLDR.
> 
> See also [Locale Class](https://docs.google.com/document/d/12ZpGW4qcNV5NLCPQHR1-viMpVs9Pod81ZkwKJvyRfqQ/edit?ts=5bebde67#heading=h.mszsvv7qat7p).
> 
> FYI There is a validity checker on the Java side, but it is not public API:
> 
> com.ibm.icu.impl.locale.LocaleValidityChecker.java
> See also com.ibm.icu.dev.test.util.TestLocaleValidity.java
> 
> _**However, you have to be very cautious in its use, as remarked above.**_

**Comment by NorbertLindenberg** [comment]
This bug report seems to assume that locales with unknown Unicode extension keys or values are rejected. Where does the spec say that?

**Comment by sffc** [comment]
The cited spec text is not really "validating" a locale; it is the logic that transforms a requested locale to a resolved locale.  If the numbering system is not found, then the `-nu-` option is simply not added to the resolved locale.

**Comment by sffc** [comment]
I agree with your assessment. At no point did I intend to suggest that there is a relationship between Intl.Locale and ResolveLocale.

**Comment by FrankYFTang** [comment]
sffc- can we then close this issue? Since 
1) the cited validation is only inside ResolveLocale and 
2) you concern is the Locale string, 
3) the ResolveLocale is not a public API, but a shared private Abstract Operation called by all Intl.* objects (except Intl.Locale itself) In other words, it is not a functions call which can be called by any JS code alone.

**Comment by sffc** [comment]
I filed the issue on behalf of Mark Davis. I agree with Frank that this issue probably ought to be closed.

I'll leave it open for a couple more days to solicit feedback from Mark, and then close it unless we get new information.

**Comment by FrankYFTang** [comment]
@macchiati ping can we close this?

**Comment by FrankYFTang** [comment]
1. ResolveLocale will  ignore any subtag that the implementation does not know about.
2. "ar-SA-u-aa-bbb" will get a number format as "ar-SA" since NumberFormat only pass in {"nu" } as as relevantExtensionKeys so the ResolveLocale will skip "aa" or any key not "nu"
3. "ar-SA-u-aa-bbb-nu-latn" will create NumberFormat with "ar-SA-u-nu-latn"
 "ar-SA-u-aa-bbb-nu-native" will create NumberFormat with "ar-SA" ("native", "traditio", or "finance" are not allowed by ECMA 402 for "nu" - see https://ecma-international.org/ecma-402/#sec-intl.numberformat-internal-slots )
4. for  "gd-SU" , that is a different story, the ResolveLocale will first call BestAvailableLocale
https://ecma-international.org/ecma-402/#sec-bestavailablelocale
Basically, it will fallback language-script-region to language-script and language in matching for each item in the passed in requestedLocale. So it will first try to see does available locale has "gd-SU", if not, it will try to see does available locale includes "gd".  For example if user call new Intl.NumberFormat(["gd-SU", "nl-SU"]) then it may create NumberFormat as "gd-SU", "gd", "nl-SU", "nl" and  the default locale, depending on which one is available in the system. If there are no "gd-SU" locale (returned by  icu::NumberFormat::getAvailableLocales()) for NumberFormat it will then use "gd" if it is available, if not, then "nl-SU", ....

**Comment by macchiati** [comment]
> ResolveLocale will ignore any subtag that the implementation does not
know about.

Since you say that's true, I have no problem with closing the ticket.

Mark
